include/internal/ktls.h

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 1995-2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef HEADER_INTERNAL_KTLS
+# define HEADER_INTERNAL_KTLS
+
+#include <sys/socket.h>
+#include <linux/version.h>
+
+#define K_MAJ   4
+#define K_MIN1  13
+#define K_MIN2  0
+#if LINUX_VERSION_CODE < KERNEL_VERSION(K_MAJ, K_MIN1, K_MIN2)
+
+# ifndef PEDANTIC
+#  warning "KTLS requires Kernel Headers >= 4.13.0"
+#  warning "Skipping Compilation of KTLS data path"
+# endif
+
+#define TLS_TX                  1       /* Set transmit parameters */
+
+#define TLS_CIPHER_AES_GCM_128                          51
+#define TLS_CIPHER_AES_GCM_128_IV_SIZE                  8
+#define TLS_CIPHER_AES_GCM_128_KEY_SIZE         16
+#define TLS_CIPHER_AES_GCM_128_SALT_SIZE                4
+#define TLS_CIPHER_AES_GCM_128_TAG_SIZE         16
+#define TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE             8
+
+#define TLS_SET_RECORD_TYPE     1
+
+struct tls_crypto_info {
+    unsigned short version;
+    unsigned short cipher_type;
+};
+
+struct tls12_crypto_info_aes_gcm_128 {
+    struct tls_crypto_info info;
+    unsigned char iv[TLS_CIPHER_AES_GCM_128_IV_SIZE];
+    unsigned char key[TLS_CIPHER_AES_GCM_128_KEY_SIZE];
+    unsigned char salt[TLS_CIPHER_AES_GCM_128_SALT_SIZE];
+    unsigned char rec_seq[TLS_CIPHER_AES_GCM_128_REC_SEQ_SIZE];
+};
+
+
+/* Dummy functions here */
+inline int ktls_enable(int fd) {
+    return -1;
+}
+
+static ossl_inline int ktls_set_offload_tx(int fd,
+                          struct tls12_crypto_info_aes_gcm_128 *crypto_info,
+                          size_t len) {
+    return -1;
+}
+
+static ossl_inline int ktls_send_ctrl_message(int fd, unsigned char record_type,
+        const void *data, size_t length) {
+    return -1;
+}
+
+#else
+
+#include <netinet/tcp.h>
+#include <linux/tls.h>
+#include <linux/socket.h>
+
+# ifndef SOL_ALG
+#  define SOL_TLS 282
+# endif
+
+# ifndef TCP_ULP
+#  define TCP_ULP 31
+# endif
+
+static ossl_inline int ktls_enable(int fd) {
+    return setsockopt(fd, SOL_TCP, TCP_ULP, "tls", sizeof("tls"));
+}
+
+static ossl_inline int ktls_set_offload_tx(int fd,
+                          struct tls12_crypto_info_aes_gcm_128 *crypto_info,
+                          size_t len) {
+    return setsockopt(fd, SOL_TLS, TLS_TX, crypto_info,
+                      sizeof(*crypto_info));
+}
+
+static ossl_inline int ktls_send_ctrl_message(int fd, unsigned char record_type,
+        const void *data, size_t length)
+{
+    struct msghdr msg = {0};
+    int cmsg_len = sizeof(record_type);
```

Comment:
side-note, but isn't sizeof [unsigned]char 1 by definition?

Reply:
Yes. I think that using record_type here helps the reader follow the code. Would you like to replace it with:
int cmsg_len = sizeof(unsigned char);
?

Reply:
That's a good point, leave it.